pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1121 @@
+	fn update_hub_asset_liquidity(
+		delta_amount: &BalanceUpdate<T::Balance>,
+		issuance_update: HubAssetIssuanceUpdate,
+	) -> DispatchResult {
+		<HubAssetLiquidity<T>>::try_mutate(|liquidity| -> DispatchResult {
+			match delta_amount {
+				BalanceUpdate::Increase(amount) => {
+					*liquidity = liquidity.checked_add(amount).ok_or(Error::<T>::Overflow)?;
+				}
+				BalanceUpdate::Decrease(amount) => {
+					*liquidity = liquidity.checked_sub(amount).ok_or(Error::<T>::Overflow)?;
+				}
+			}
+			Ok(())
+		})?;
+
+		if issuance_update == HubAssetIssuanceUpdate::AdjustSupply {
+			match delta_amount {
+				BalanceUpdate::Increase(amount) => {
+					T::Currency::deposit(T::HubAssetId::get(), &Self::protocol_account(), *amount)?;
+				}
+				BalanceUpdate::Decrease(amount) => {
+					T::Currency::withdraw(T::HubAssetId::get(), &Self::protocol_account(), *amount)?;
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Update imbalance with given delta_imbalance and write new value to storage.
+	fn update_imbalance(
+		current_imbalance: SimpleImbalance<T::Balance>,
+		delta_imbalance: BalanceUpdate<T::Balance>,
+	) -> DispatchResult {
+		let imbalance = match delta_imbalance {
+			BalanceUpdate::Decrease(amount) => current_imbalance.sub(amount).ok_or(Error::<T>::Overflow)?,
+			BalanceUpdate::Increase(amount) => current_imbalance.add(amount).ok_or(Error::<T>::Overflow)?,
+		};
+		<HubAssetImbalance<T>>::put(imbalance);
+
+		Ok(())
+	}
+
+	/// Recalculate imbalance and call update_imbalance with new imbalance value.
+	fn recalculate_imbalance(
+		asset_state: &AssetState<T::Balance>,
+		delta_amount: BalanceUpdate<T::Balance>,
+	) -> DispatchResult {
+		let current_imbalance = <HubAssetImbalance<T>>::get();
+		let current_hub_asset_liquidity = <HubAssetLiquidity<T>>::get();
+
+		if current_imbalance.value != T::Balance::zero() && current_hub_asset_liquidity != T::Balance::zero() {
+			// if any is 0, the delta is 0 too.
+
+			let p1 = FixedU128::from((asset_state.hub_reserve, asset_state.reserve));
+			let p2 = FixedU128::from((current_imbalance.value, current_hub_asset_liquidity));
+			let p3 = p1.checked_mul(&p2).ok_or(Error::<T>::Overflow)?;
+
+			let delta_imbalance = p3.checked_mul_int(*delta_amount).ok_or(Error::<T>::Overflow)?;
+
+			match delta_amount {
+				BalanceUpdate::Increase(_) => {
+					Self::update_imbalance(current_imbalance, BalanceUpdate::Increase(delta_imbalance))
+				}
+				BalanceUpdate::Decrease(_) => {
+					Self::update_imbalance(current_imbalance, BalanceUpdate::Decrease(delta_imbalance))
+				}
+			}
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Update total tvl and asset tvl. Write new total tvl value to storage, and mutate asset tvl accordingly.
+	fn update_tvl(asset_state: &mut AssetState<T::Balance>) -> DispatchResult {
+		let (stable_asset_reserve, stable_asset_hub_reserve) = Self::stable_asset()?;
+
+		if stable_asset_reserve != T::Balance::zero() && stable_asset_hub_reserve != T::Balance::zero() {
```

Comment:
Well, 

`FixedU128::from( (100,0) )`  panics. 

So, it is also because of that. 

now that i think of it - i use this `from( (N,D) )` in other places too, without explicitly checking that the D != 0. 

What do you think ?

Reply:
I think it makes sense to use [checked_from_rational](https://github.com/paritytech/substrate/blob/4d3d480d8149fbca2834d05cc050a01f9a25afbf/primitives/arithmetic/src/fixed_point.rs#L142) whenever the denominator could be zero and then return `ArithmeticError::DivisionByZero`